connectors/sources/sharepoint_online.py

```diff
@@ -0,0 +1,740 @@
+        """Fetch API token for usage with Graph API
+
+        Returns:
+            (str, int) - a tuple containing access token as a string and number of seconds it will be valid for as an integer"""
+
+        url = f"{GRAPH_API_AUTH_URL}/{self._tenant_id}/oauth2/v2.0/token"
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+        data = f"client_id={self._client_id}&scope=https://graph.microsoft.com/.default&client_secret={self._client_secret}&grant_type=client_credentials"
+
+        async with self._http_session.post(url, headers=headers, data=data) as resp:
+            json_response = await resp.json()
+            access_token = json_response["access_token"]
+            expires_in = int(json_response["expires_in"])
+            return access_token, expires_in
+
+
+class SharepointRestAPIToken(MicrosoftSecurityToken):
+    """Token to connect to Sharepoint REST API endpoints."""
+
+    async def _fetch_token(self):
+        """Fetch API token for usage with Sharepoint REST API
+
+        Returns:
+            (str, int) - a tuple containing access token as a string and number of seconds it will be valid for as an integer"""
+
+        url = f"{REST_API_AUTH_URL}/{self._tenant_id}/tokens/OAuth/2"
+        # GUID in resource is always a constant used to create access token
+        data = {
+            "grant_type": "client_credentials",
+            "resource": f"00000003-0000-0ff1-ce00-000000000000/{self._tenant_name}.sharepoint.com@{self._tenant_id}",
+            "client_id": f"{self._client_id}@{self._tenant_id}",
+            "client_secret": self._client_secret,
+        }
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+
+        async with self._http_session.post(url, headers=headers, data=data) as resp:
+            json_response = await resp.json()
+            access_token = json_response["access_token"]
+            expires_in = int(json_response["expires_in"])
+
+            return access_token, expires_in
+
+
+class PermissionsMissing(Exception):
+    """Exception class to notify that specific Application Permission is missing for the credentials used.
+    See: https://learn.microsoft.com/en-us/graph/permissions-reference
+    """
+    pass
+
+
+class MicrosoftAPISession:
+    def __init__(self, http_session, api_token, scroll_field):
+        self._http_session = http_session
+        self._api_token = api_token
+        self._semaphore = asyncio.Semaphore(
+            10
+        )  # TODO: make configurable, that's a scary property
+
+        # Graph API and Sharepoint API scroll over slightly different fields:
+        # - odata.nextPage for Sharepoint REST API uses
+        # - @odata.nextPage for Graph API uses - notice the @ glyph
+        # Therefore for flexibility I made it a field passed in the initializer,
+        # but this abstraction can be better.
+        self._scroll_field = scroll_field
+
+    async def fetch(self, url, debug=False):
```

Comment:
Why is `debug` used here?

Reply:
Cause it's a WIP and I sometimes push things that won't go into production :D